jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlParameterDecoder.java

```diff
@@ -0,0 +1,330 @@
+        this.maxKeys = maxKeys;
+        this.allowBadEncoding = allowBadEncoding;
+        this.allowBadPercent = allowBadPercent;
+        this.allowTruncatedEncoding = allowTruncatedEncoding;
+        this.allowPartialBufferString = allowBadEncoding;
+    }
+
+    /**
+     * <p>Parse a String completely.</p>
+     *
+     * <p>The {@code newFieldAdder} is called for each encountered {@code key=value} pair.</p>
+     *
+     * @param str the string to parse, completing the parsing after parsing.
+     * @return true if there were no coding errors, false otherwise.
+     * @throws CharacterCodingException if a coding issue is encountered with the
+     * provided {@link CharsetStringBuilder} and the specific condition
+     * is not allowed by one of the {@code allow*} parameters on the constructor.
+     */
+    public boolean parse(String str) throws CharacterCodingException
+    {
+        return parse(str, 0, str.length());
+    }
+
+    /**
+     * <p>Parse a String completely.</p>
+     *
+     * <p>The {@code newFieldAdder} is called for each encountered {@code key=value} pair.</p>
+     *
+     * @param str the string to parse, completing the parsing after parsing.
+     * @param offset the offset in the string to start parsing from.
+     * @param length the length of the substring to parse.
+     * @return true if there were no coding errors, false otherwise.
+     * @throws CharacterCodingException if a coding issue is encountered with the
+     * provided {@link CharsetStringBuilder} and the specific condition
+     * is not allowed by one of the {@code allow*} parameters on the constructor.
+     */
+    public boolean parse(String str, int offset, int length) throws CharacterCodingException
+    {
+        int end = offset + length;
+        for (int i = offset; i < end; i++)
+        {
+            parse(str.charAt(i));
+        }
+        complete();
+        return builder.hasCodingErrors();
+    }
+
+    /**
+     * <p>Parse a InputStream completely.</p>
+     *
+     * <p>The {@code newFieldAdder} is called for each encountered {@code key=value} pair.</p>
+     *
+     * <p>The InputStream is read until EOF</p>
+     *
+     * @param input the InputStream to parse, completing the parsing after parsing.
+     * @return true if there were no coding errors, false otherwise.
+     * @throws CharacterCodingException if a coding issue is encountered with the
+     * provided {@link CharsetStringBuilder} and the specific condition
+     * is not allowed by one of the {@code allow*} parameters on the constructor.
+     */
+    public boolean parse(InputStream input, Charset charset) throws IOException
+    {
+        int c;
+        InputStreamReader reader = new InputStreamReader(input, charset);
+        while ((c = reader.read()) != -1)
+        {
+            parse((char)c);
+        }
+        complete();
+        return builder.hasCodingErrors();
+    }
+
+    private void parse(char c) throws CharacterCodingException
```

Comment:
I think this method is too large to be JIT inlined.    So rather than call it in a loop char by char, perhaps it would be better to pass in a `Supplier<Character>` or an interface with a `char` equivalent.     The loop could then be in the method and bounds checking could be common.

@lorban is my optimization knowledge out of date? will this make a difference?

Reply:
My knowledge of this level of detail probably is out of date too, but from the top of my head, all methods under ~20 bytecodes are candidates to be inlined.

This method seems to fit that bytecode limit, and even if it didn't, I fail to see how an equivalent method accepting a `Supplier<Character>` would be that different.

If you want to be certain, writing a microbenchmark is the way to go.